lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,447 @@
+        Http http = new Http(new ClientOptions(), null);
+        Http.RequestBody requestBody = new Http.ByteArrayRequestBody(new byte[0], NanoHTTPD.MIME_PLAINTEXT);
+        URL url = new URL("http://localhost:" + server.getListeningPort() + "/status/501");
+
+        /* Verify
+         *   - HostFailedException is thrown */
+        thrown.expect(Http.HostFailedException.class);
+        http.httpExecute(url, Http.GET, new Param[0], requestBody, false, null);
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is performing a fallback,
+     * when api returns a response code between 500 and 504
+     * </p>
+     * <p>
+     * Spec: RSC15d
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_throws_502() throws MalformedURLException, AblyException {
+        Http http = new Http(new ClientOptions(), null);
+        Http.RequestBody requestBody = new Http.ByteArrayRequestBody(new byte[0], NanoHTTPD.MIME_PLAINTEXT);
+        URL url = new URL("http://localhost:" + server.getListeningPort() + "/status/502");
+
+        /* Verify
+         *   - HostFailedException is thrown */
+        thrown.expect(Http.HostFailedException.class);
+        http.httpExecute(url, Http.GET, new Param[0], requestBody, false, null);
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is performing a fallback,
+     * when api returns a response code between 500 and 504
+     * </p>
+     * <p>
+     * Spec: RSC15d
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_throws_503() throws MalformedURLException, AblyException {
+        Http http = new Http(new ClientOptions(), null);
+        Http.RequestBody requestBody = new Http.ByteArrayRequestBody(new byte[0], NanoHTTPD.MIME_PLAINTEXT);
+        URL url = new URL("http://localhost:" + server.getListeningPort() + "/status/503");
+
+        /* Verify
+         *   - HostFailedException is thrown */
+        thrown.expect(Http.HostFailedException.class);
+        http.httpExecute(url, Http.GET, new Param[0], requestBody, false, null);
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is performing a fallback,
+     * when api returns a response code between 500 and 504
+     * </p>
+     * <p>
+     * Spec: RSC15d
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_throws_504() throws MalformedURLException, AblyException {
+        Http http = new Http(new ClientOptions(), null);
+        Http.RequestBody requestBody = new Http.ByteArrayRequestBody(new byte[0], NanoHTTPD.MIME_PLAINTEXT);
+        URL url500 = new URL("http://localhost:" + server.getListeningPort() + "/status/504");
+
+        /* Verify
+         *   - HostFailedException is thrown */
+        thrown.expect(Http.HostFailedException.class);
+        http.httpExecute(url500, Http.GET, new Param[0], requestBody, false, null);
```

Comment:
Certainly not, I was surprised you wrote tests for all 5 50x status codes. I am saying that you have 5 test for 500-504 which could have more easily be put into a for loop, and you have no test to ensure that say a 505 does NOT result in a fallback host check, which means you are only testing the conditions you expect.